langpuppet/gapi.go

```diff
@@ -0,0 +1,290 @@
+		return cli.StringFlag{
+			Name:  "lp-" + flag.Name,
+			Value: flag.Value,
+		}
+	})
+}
+
+// Init initializes the langpuppet GAPI struct.
+func (obj *GAPI) Init(data gapi.Data) error {
+	if obj.initialized {
+		return fmt.Errorf("already initialized")
+	}
+	if obj.InputURI == "" {
+		return fmt.Errorf("the InputURI param must be specified")
+	}
+	obj.data = data // store for later
+
+	fs, err := obj.data.World.Fs(obj.InputURI) // open the remote file system
+	if err != nil {
+		return errwrap.Wrapf(err, "can't load data from file system `%s`", obj.InputURI)
+	}
+
+	dataLang := &gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(lang.Name+": "+format, v...)
+		},
+	}
+
+	dataPuppet := &gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	obj.langGAPI = &lang.GAPI{}
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	obj.puppetGAPI = &puppet.GAPI{}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	if obj.currentLangGraph == nil || obj.currentPuppetGraph == nil {
+		return nil, fmt.Errorf("%s: graph was requested before both Lang and Puppet graphs were available", Name)
+	}
+
+	lg := obj.currentLangGraph
+	pg := obj.currentPuppetGraph
+
+	g := &pgraph.Graph{
+		Name: lg.Name + "x" + pg.Name,
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		var currentLangGraph *pgraph.Graph
+		var currentPuppetGraph *pgraph.Graph
+
+		for {
+			var graph_err error
+
+			select {
+			case nextChild <- nextLang:
+				graph_err = nextChild.Err
+				if nextChild.Err == nil {
+					obj.currentLangGraph, graph_err = obj.langGAPI.Graph()
```

Comment:
Hmm... I'm not 100% sure this is the right approach. IMO (I'd have to think about it.) Next shouldn't call Graph. Keep them separate unless you think I'm wrong.

Reply:
Fixed.